### PR TITLE
SymmetricCircuits & Current opt state vector reduction

### PR DESCRIPTION
**Pull request by oliverfunk**
## Linked Issues

Closes #{ID}

## Description

Briefly, here is a proof for why the derivative values can be added when you transform a function into a new function by setting some parameters to equal to each other:

Say we have some scalar function f of two variables x and y:

image

The gradient of f is:

image

Say we now have a new function u that sets y = x:

image

Then the gradient of u can then be expressed as:

image

An example

Say we have f:

image

Then we have u:

image

One can see, if you substitute y=x into the grad f function you get grad u.

## Checklist

I confirm that I have completed the following checks:

- [ ] Tests run locally and pass `pytest tests --reactor`
- [ ] Code quality checks run locally and pass `pre-commit run --from-ref develop --to-ref HEAD`
- [ ] Documentation built locally and checked `sphinx-build -W documentation/source documentation/build`

### CODE REVIEW

**Comment by oliverfunk** [comment]
This is nearly ready.

I need to redo the `CoilSetSymmetryStatus` feature, as will not work properly if one uses a `Circuit` instead of Sym. Circuit.

File: bluemira/equilibria/coils/_grouping.py [hunk 1]
```diff
@@ -1184,10 +1184,143 @@ def set_optimisation_state(
...
+            for c in self._coils
+        ]
+        return [*flatten_iterable(optimisable_coil_names)]
+
+    @property
+    def all_current_optimisable_coils(self) -> list[Coil]:
+        """
+        Get the names of all coils that can be current optimised.
+        """
+        return [self[cn] for cn in self.current_optimisable_coil_names]
```
Comment (geograham) [inline bluemira/equilibria/coils/_grouping.py:1214]: """ Get list of all coils that can be current optimised."""

File: bluemira/equilibria/coils/_grouping.py [hunk 2]
```diff
@@ -1184,10 +1184,143 @@ def set_optimisation_state(
...
+
+    def get_current_optimisable_coils(
+        self, coil_names: list[str] | None = None
+    ) -> list[Coil]:
+        """
+        Get the coils that can be current optimised.
```
Comment (geograham) [inline bluemira/equilibria/coils/_grouping.py:1220]: Should this be changed to something like 'get list of selected coils from coils that can be current optimised'?

File: bluemira/equilibria/optimisation/problem/base.py [hunk 3]
```diff
@@ -189,26 +199,58 @@ def update_magnetic_constraints(
...
+
             d: ConstraintT = {
                 "name": f_constraint.name,
-                "f_constraint": f_constraint.f_constraint,
-                "df_constraint": getattr(f_constraint, "df_constraint", None),
+                "f_constraint": f_c,
+                "df_constraint": df_c,
                 "tolerance": constraint.tolerance,
             }
             # TODO: tidy this up, so the interface guarantees this works!
```
Comment (geograham) [inline bluemira/equilibria/optimisation/problem/base.py:248]: Just wondering what this #TODO is from? Do we still need it?

**Comment by geograham** [comment]
Hello, looks good, I think that you should also update the 'Circuits' section of the docs. Maybe with a note on what a a fully, partial and non-symmetric circuit is, a diagram, and the f and df modifications?

File: bluemira/equilibria/optimisation/problem/base.py (hunk 3 above)
Comment (oliverfunk) [reply bluemira/equilibria/optimisation/problem/base.py:248]: Yeah.. It's because the ABC doesn't have a good interface. It should all be improved.

**Review by geograham — DISMISSED** [review]

**Review by je-cook — APPROVED** [review]
Congrats!

**Pull request merged**